cache_req: Keep the files provider as the first domain to be searched

fidencio opened the pull request:
Currently we can't guarantee any order on which domain will the first to
be searched. More than that, in case domain_resolution_order is set, we
actually enforce that the first domain searched will respect the option
set.

This behaviour is not exactly the expect, as the implicit files domain
has to be searched first in order to avoid querying for local users in
remote domains. In order to enforce this, let's just keep the files
domain as the first to be searched, always!

Resolves:
https://pagure.io/SSSD/sssd/issue/3768

Signed-off-by: Fabiano Fidêncio <fidencio@redhat.com>

sumit-bose commented:
Not strictly related but in confdb.c strcasecmp() is used to check for the files provider. Maybe it would make sense to add a macro like e.g. IS_FILES_PROVIDER() (and fix the other occurrences in a different PR)  to have a consistent check?

fidencio commented:
retest this, please

fidencio commented:
@sumit-bose, I've done that as part of https://github.com/SSSD/sssd/pull/607.
Take a look at https://github.com/SSSD/sssd/pull/607/commits/768039bd1adda130b550f9e405bac6d320858662 ...

The plan is to rebase this one atop of the 607 (or vice-versa, depending on what will be acked first)

fidencio commented:
@sumit-bose, btw, I've updated my other patch to also use strcasecmp() instead of strcmp(), thanks.

fidencio commented:
@sumit-bose, I've updated the patchset.

Changes since last version:
- rebased atop of git master;
- making use of is_files_provider() call that was pushed Today;
- wrote tests for the changes in this patchset;
- wrote tests for another issue fixed some time ago (https://pagure.io/SSSD/sssd/issue/3743);

fidencio commented:
And make-check-valgrind is failing. Let me add the "Changes Requested" label again.

sumit-bose commented on src/tests/domain_resolution_order-tests.c:
```diff
@@ -0,0 +1,227 @@
+*/
+
+#include <stdlib.h>
+#include <check.h>
```
Is there a reason for using the old check test-framework and not cmocka?

fidencio replied:
Well, no. No specific reason.

The few things that crossed my mind while writing those is that:
- We don't need any mock to run those tests;
- We don't run cmocka tests in a few platforms (EL)

Would you prefer that I migrate those to cmocka even if I don't use anything from there?

sumit-bose replied:
I think it is a fair point that we do not run the cmocka tests on all platforms because cmocka is not available, so please keep the test as it is.

sumit-bose commented on src/tests/domain_resolution_order-tests.c:
```diff
@@ -0,0 +1,227 @@
+
+        dom->name = discard_const(domains[i]);
+        dom->provider = discard_const(providers[i]);
+        dom->conn_name = discard_const("conn");
```
I think you have to copy here because the values might go out-of-scope.

fidencio commented:
@sumit-bose, I've updated the patchset.

fidencio commented:
After a discussion on IRC we've agreed that would be better to have the tests using cmocka instead of check framework. Thus, the patchset has been updated again.

sumit-bose commented:
Thank you, I'll run some test before giving ACK.

sumit-bose commented:
Ok, test went well, Coverity and CI passed, ACK.
